internal/metamorph/store/postgresql/postgres.go

```diff
@@ -573,6 +591,135 @@ func (p *PostgreSQL) UpdateStatusBulk(ctx context.Context, updates []store.Updat
+			) AS bulk_query
+			WHERE metamorph.transactions.hash=bulk_query.hash
+				AND metamorph.transactions.status <= bulk_query.status
+				AND LENGTH(metamorph.transactions.competing_txs) < LENGTH(bulk_query.competing_txs)
+		RETURNING metamorph.transactions.stored_at
+		,metamorph.transactions.announced_at
+		,metamorph.transactions.mined_at
+		,metamorph.transactions.hash
+		,metamorph.transactions.status
+		,metamorph.transactions.block_height
+		,metamorph.transactions.block_hash
+		,metamorph.transactions.callback_url
+		,metamorph.transactions.callback_token
+		,metamorph.transactions.full_status_updates
+		,metamorph.transactions.reject_reason
+		,metamorph.transactions.competing_txs
+		,metamorph.transactions.raw_tx
+		,metamorph.transactions.locked_by
+		,metamorph.transactions.merkle_path
+		,metamorph.transactions.retries
+		;
+    `
+
+	txHashes := make([][]byte, len(updates))
+	for i, update := range updates {
+		txHashes[i] = update.Hash.CloneBytes()
+	}
+
+	tx, err := p.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get current competing transactions and lock them for update
+	rows, err := tx.QueryContext(ctx, `SELECT hash, competing_txs FROM metamorph.transactions WHERE hash in (SELECT UNNEST($1::BYTEA[])) FOR UPDATE`, pq.Array(txHashes))
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			return nil, err
+		}
+		return nil, err
+	}
+	defer rows.Close()
+
+	dbData := make([]*store.StoreData, 0)
+	for rows.Next() {
+		data := &store.StoreData{}
+
+		var hash []byte
+		var competingTxs string
+
+		err := rows.Scan(
+			&hash,
+			&competingTxs,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(hash) > 0 {
+			data.Hash, err = chainhash.NewHash(hash)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if competingTxs != "" {
+			data.CompetingTxs = strings.Split(competingTxs, ",")
+		}
+
+		dbData = append(dbData, data)
+	}
+
+	statuses := make([]metamorph_api.Status, len(updates))
+	competingTxs := make([]string, len(updates))
+	rejectReasons := make([]string, len(updates))
+
+	for i, update := range updates {
+		txHashes[i] = update.Hash.CloneBytes()
+		statuses[i] = update.Status
+
+		rejectReasons[i] = ""
+		if update.Error != nil {
+			rejectReasons[i] = update.Error.Error()
+		}
+
+		for _, data := range dbData {
+			if update.Hash.IsEqual(data.Hash) {
+				// get unique values from merged existing competing txs
+				// and incoming ones to avoid duplicates or overridings
+				uniqueTxs := mergeUnique(update.CompetingTxs, data.CompetingTxs)
+				competingTxs[i] = strings.Join(uniqueTxs, ",")
+				break
+			}
+		}
+	}
+
+	rows, err = tx.QueryContext(ctx, qBulk, pq.Array(txHashes), pq.Array(statuses), pq.Array(rejectReasons), pq.Array(competingTxs))
+	if err != nil {
+		if err := tx.Rollback(); err != nil {
+			return nil, err
+		}
+		return nil, err
+	}
+	defer rows.Close()
+
+	res, err := p.getStoreDataFromRows(rows)
+	if err != nil {
+		return res, err
+	}
+
+	err = tx.Commit()
```

Comment:
Shouldn't we commit the transaction before we map the rows?

Reply:
That's a good question and I was thinking about it a lot. 

I decided to rollback the transaction in case of error when mapping rows for 2 reasons:
1. It's consistent with other methods in this file.
2. In case something malformed is added to db and it throws an error when trying to map it from rows, we probably don't want that malformed record in the db anyways, because it will always return an error when retrieving it.

I implemented the rollback, but if you think we should do it another way, let me know